logger_bench_test.go

```diff
@@ -238,3 +240,130 @@ func Benchmark100Fields(b *testing.B) {
 		logger.With(first...).Info("Child loggers with lots of context.", second...)
 	}
 }
+
+func dummy(wg *sync.WaitGroup, s string, i int) string {
+	if i == 0 {
+		wg.Wait()
+		return "1" + s
+	}
+	return dummy(wg, s, i-1)
+}
+
+func stackGrower(n int) *sync.WaitGroup {
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+
+	go dummy(&wg, "hi", n)
+	return &wg
+}
+
+func BenchmarkAny(b *testing.B) {
+	var (
+		logger = New(
+			zapcore.NewCore(
+				zapcore.NewJSONEncoder(NewProductionConfig().EncoderConfig),
+				&ztest.Discarder{},
+				DebugLevel,
+			),
+		)
+		key = "some-long-string-longer-than-16"
+	)
+
+	tests := []struct {
+		name   string
+		typed  func() Field
+		anyArg any
+	}{
+		{
+			name:   "string",
+			typed:  func() Field { return String(key, "yet-another-long-string") },
+			anyArg: "yet-another-long-string",
+		},
+		{
+			name:   "stringer",
+			typed:  func() Field { return Stringer(key, InfoLevel) },
+			anyArg: InfoLevel,
+		},
+	}
+
+	for _, tt := range tests {
+		b.Run(tt.name+"-typ-no-logger", func(b *testing.B) {
+			for i := 0; i < b.N; i++ {
+				f := tt.typed()
+				runtime.KeepAlive(f)
+			}
+		})
+		b.Run(tt.name+"-any-no-logger", func(b *testing.B) {
+			for i := 0; i < b.N; i++ {
+				f := Any(key, tt.anyArg)
+				runtime.KeepAlive(f)
+			}
+		})
+		b.Run(tt.name+"-typ-logger", func(b *testing.B) {
+			for i := 0; i < b.N; i++ {
+				logger.Info("", tt.typed())
+			}
+		})
+		b.Run(tt.name+"-any-logger", func(b *testing.B) {
+			for i := 0; i < b.N; i++ {
+				logger.Info("", Any(key, tt.anyArg))
+			}
+		})
+		b.Run(tt.name+"-typ-logger-go", func(b *testing.B) {
+			wg := sync.WaitGroup{}
+			wg.Add(b.N)
+			b.ResetTimer()
+			for i := 0; i < b.N; i++ {
+				go func() {
+					logger.Info("", tt.typed())
+					wg.Done()
+				}()
+			}
+			wg.Wait()
```

Comment:
Actually, is this right? It'll spawn `b.N` goroutines and count that as part of a run.
I think maybe you want a fixed number of goroutines spawned inside each b.N loop, and wait for them to finish also in the loop?

```
const Count = 1000
for i := 0; i < b.N; i++ {
    wg.Add(Concurrency)
    for j := 0; j < Count; j++ {
      go func() { ... }()
    }
    wg.Wait()
}
```

Reply:
So the actual issue, as (badly) reported by a user was: "if I start a new goroutine and `zap.Any` in it, it's slow" - this is where this came from, with an exact translation.

The goal was also to have sequential progress:
- how long does it take to create a `Field`
- how long does it take to log a `Field`
- how long does it take to log a `Field` in a new goroutine - including the setup and teardown cost.

Spawning o 1000 goroutines would at least break the sequentially above, I think?

It seemed that as long as we're comparing the same cost for typed version and `Any`, the benchmarks are representative.